internal/go_repository.bzl

```diff
@@ -286,6 +333,28 @@ def _go_repository_impl(ctx):
     # Override external GO111MODULE, because it is needed by module mode, no-op in repository mode
     fetch_repo_env["GO111MODULE"] = "on"
 
+    if can_use_direct_download:
+        module = ctx.attr.replace if ctx.attr.replace else ctx.attr.importpath
+        proxy_url = _get_goproxy_url_for_module(env, module)
+
+        if proxy_url:
+            result = ctx.download_and_extract(
+                url = "%s/%s/@v/%s.zip" % (
+                    proxy_url,
+                    _escape_go_module_path(module),
+                    ctx.attr.version,
+                ),
+                sha256 = ctx.attr.sha256,
+                canonical_id = ctx.attr.canonical_id,
+                stripPrefix = module + "@" + ctx.attr.version,
+                type = ctx.attr.type,
+                auth = use_netrc(read_user_netrc(ctx), ctx.attr.urls, ctx.attr.auth_patterns),
+                allow_fail = True,
+            )
+            if result.success:
```

Comment:
In which cases do we expect this to fail? Should we print a warning if it does or fail instead and provide a way to opt into the fallback?

Reply:
ah, good to know. That reaffirms my belief that if we go down this route, we should not try to replicate the full fallback logic, but defer to `go mod download` after the first failure, which is what this PR does.

Do you think there are cases where ctx.download_and_extract does NOT fail but we should still do a fallback?